rclrs/src/executor.rs

```diff
@@ -1,84 +1,466 @@
+    channel: Arc<dyn WorkerChannel>,
+    wakeup_wait_set: Arc<GuardCondition>,
+}
+
+impl WorkerCommands {
+    pub(crate) fn add_to_wait_set(&self, waitable: Waitable) {
+        self.channel.add_to_waitset(waitable);
+    }
+
+    pub(crate) fn run_async<F>(&self, f: F)
+    where
+        F: 'static + Future<Output = ()> + Send,
+    {
+        self.channel.add_async_task(Box::pin(f));
+    }
+
+    pub(crate) fn run_on_payload(&self, task: PayloadTask) {
+        self.channel.send_payload_task(task);
+    }
+
+    pub(crate) fn add_activity_listener(&self, listener: WeakActivityListener) {
+        self.channel.add_activity_listener(listener);
+    }
+
+    /// Get a guard condition that can be used to wake up the wait set of the executor.
+    pub(crate) fn get_guard_condition(&self) -> &Arc<GuardCondition> {
+        &self.wakeup_wait_set
+    }
+}
+
+/// Channel to send commands related to a specific worker.
+pub trait WorkerChannel: Send + Sync {
+    /// Add a new item for the executor to run.
+    fn add_async_task(&self, f: BoxFuture<'static, ()>);
+
+    /// Add new entities to the waitset of the executor.
+    fn add_to_waitset(&self, new_entity: Waitable);
+
+    /// Send a one-time task for the worker to run with its payload.
+    fn send_payload_task(&self, f: PayloadTask);
+
+    /// Send something to listen to worker activity.
+    fn add_activity_listener(&self, listener: WeakActivityListener);
+}
+
+/// Encapsulates a task that can operate on the payload of a worker
+pub type PayloadTask = Box<dyn FnOnce(&mut dyn Any) + Send>;
+
+/// This is constructed by [`ExecutorCommands`] and passed to the [`ExecutorRuntime`]
+/// to create a new worker. Downstream users of rclrs should not be using this class
+/// unless you are implementing your own [`ExecutorRuntime`].
+pub struct ExecutorWorkerOptions {
+    /// The context that the executor is associated with
+    pub context: Context,
+    /// The payload that the worker provides to different primitives.
+    pub payload: Box<dyn Any + Send>,
+    /// The guard condition that should wake up the wait set for the worker.
+    pub guard_condition: Arc<GuardCondition>,
+}
+
+/// This trait defines the interface for passing new items into an executor to
+/// run.
+pub trait ExecutorChannel: Send + Sync {
+    /// Create a new channel specific to a worker whose payload must be
+    /// initialized with the given function.
+    fn create_worker(
+        &self,
+        options: ExecutorWorkerOptions,
+    ) -> Arc<dyn WorkerChannel>;
+
+    /// Wake all the wait sets that are being managed by this executor. This is
+    /// used to make sure they respond to [`ExecutorCommands::halt_spinning`].
+    fn wake_all_wait_sets(&self);
+}
+
+/// This trait defines the interface for having an executor run.
+pub trait ExecutorRuntime: Send {
+    /// Get a channel that can add new items for the executor to run.
+    fn channel(&self) -> Arc<dyn ExecutorChannel>;
```

Comment:
Saw this while reviewing #428, couldn't we use associated types to get away from trait objects here? I don't think we really need the extra flexibility. Do we really plan on doing something like storing executors in a heterogeneous container? (i.e. `Vec<Box<dyn ExecutorRuntime>>`)
```rust
pub trait ExecutorRuntime: Send {
    type ExecutorChannel: ExecutorChannel + Send + Sync;

    fn channel(&self) -> Arc<Self::ExecutorChannel>;

    fn spin(&mut self, conditions: SpinConditions) -> Vec<RclrsError>;

    fn spin_async(
        self,
        conditions: SpinConditions,
    ) -> BoxFuture<'static, (Self, Vec<RclrsError>)>
    where
        Self: Sized;
}
```
As long as Self is Sized, I think we can also remove the extra `Box` in the future return

Reply:
I did consider exactly what you're describing, but the `dyn ExecutorChannel` gets used inside [`ExecutorCommands`](https://github.com/mxgrey/ros2_rust/blob/6d3f7e4660cf6c82e2416139829331c9b1566b0c/rclrs/src/executor.rs#L124) which is held by many different objects including [`Node`](https://github.com/mxgrey/ros2_rust/blob/6d3f7e4660cf6c82e2416139829331c9b1566b0c/rclrs/src/node.rs#L98), [`SubscriptionExecutable`](https://github.com/mxgrey/ros2_rust/blob/6d3f7e4660cf6c82e2416139829331c9b1566b0c/rclrs/src/subscription.rs#L215), [`ServiceExecutable`](https://github.com/mxgrey/ros2_rust/blob/6d3f7e4660cf6c82e2416139829331c9b1566b0c/rclrs/src/service.rs#L208), and in the future `Action` will probably need it.

If we used an associated type, then we would need to add a generic argument for the type of the `ExecutorRuntime` to `ExecutorCommands`, which would then require us to add a generic argument for the `ExecutorRuntime` to `Node`, `SubscriptionExecutable`, `ServiceExecutable`, and `Client`. That would undermine the separation of concerns that we currently enjoy between application code that creates nodes and primitives vs application code that chooses the runtime.

With what you're proposing, to be agnostic to the choice of runtime, user code would need to be like

```rust
struct MyNodeData<E: ExecutorRuntime> {
    node: Node<E>,
    subscription: Subscription<Msg>,
    client: Client<E, Srv>,
}

fn setup_my_application<E: ExecutorRuntime>(executor: Executor<E>) -> MyNode<E> {
    let node = executor.create_node("my_node");
    let subscription = node.create_subscription("my_topic", do_something);
    let client = node.create_client("my_service");
    MyNode { node, subscription, client }
}
```

Probably the worst part of this IMO is that we'll see inconsistencies between what kind of generic parameters are needed by Node vs Subscription vs Client vs Service. And if we change some implementation detail of Subscription or Service, it's possible that we'll need to add a generic parameter for the runtime to them.

On the other hand, using a trait object gives us a clean separation between these concerns, allows all the primitives to have consistent generic parameters, and reduces the coupling between the user-facing API and the implementation details.